pkg/kubelet/certificate/kubelet_test.go

```diff
@@ -100,3 +105,123 @@ func TestAddressesToHostnamesAndIPs(t *testing.T) {
 		})
 	}
 }
+
+func atomicWriteFile(name string, data []byte, perm os.FileMode) error {
+	tmp := name + ".tmp"
+	if err := os.WriteFile(tmp, data, perm); err != nil {
+		return err
+	}
+	return os.Rename(tmp, name)
+}
+
+func removeThenCreate(name string, data []byte, perm os.FileMode) error {
+	if err := os.Remove(name); err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	}
+	return os.WriteFile(name, data, perm)
+}
+
+func createCertAndKeyFiles(certPath, keyPath string, useRename bool) error {
+	cert, key, err := cert.GenerateSelfSignedCertKey("k8s.io", nil, nil)
+	if err != nil {
+		return err
+	}
+
+	createFile := removeThenCreate
+	if useRename {
+		createFile = atomicWriteFile
+	}
+
+	if err := createFile(certPath, cert, os.FileMode(0644)); err != nil {
+		return err
+	}
+
+	if err := createFile(keyPath, key, os.FileMode(0600)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func TestKubeletServerCertificateFromFiles(t *testing.T) {
+	tests := []struct {
+		name           string
+		useAtomicWrite bool
+	}{
+		{
+			name:           "use atomic write",
+			useAtomicWrite: true,
+		},
+		{
+			name:           "remove and create",
+			useAtomicWrite: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			certDir := t.TempDir()
+			rotateCertErrs := make(chan error, 10)
+			defer func() {
+				if len(rotateCertErrs) != 0 {
+					t.Errorf("got errors when rotating certificate files in the test")
+				}
+				close(rotateCertErrs)
+			}()
+
+			certPath := filepath.Join(certDir, "kubelet.cert")
+			keyPath := filepath.Join(certDir, "kubelet.key")
+
+			err := createCertAndKeyFiles(certPath, keyPath, tt.useAtomicWrite)
+			if err != nil {
+				t.Fatalf("Unable to setup cert files: %v", err)
+			}
+
+			// simulate certificate files update in the background
+			go func() {
+				time.Sleep(1 * time.Second)
+				if err := createCertAndKeyFiles(certPath, keyPath, tt.useAtomicWrite); err != nil {
```

Comment:
I think it would be simpler to just do this inline after the first call to `m.Current`.  That avoids having any race in the test.